### ISSUE TITLE
RANSAC plane estimation does not work with dense point clouds

### ISSUE THREAD

**Issue by themightyoarfish** [PointCloudLibrary/pcl#1880]
## Your Environment
* Operating System and version: macOS 10.12.5
* Compiler: clang
* PCL Version: 1.7

## Expected Behavior
Plane fitting via RANSAC works independently of point density

## Current Behavior
Estimation works when every other point is discarded, but fails miserably on the full 852360 points

## Code to Reproduce
I have uploaded the cloud and code here: https://github.com/themightyoarfish/hyperspectral_calibration_testbed/tree/pcl_bug
It is runnable via `./pcd_test <cloud> <factor>` where factor is the downsampling factor. Passing 1 results in

![screen shot 2017-06-01 at 15 06 48](https://cloud.githubusercontent.com/assets/11613312/26681070/140b2d0c-46dc-11e7-8961-6648f0ca82b0.png)

whereas passing 2 (ie discarding every other point) results in

![screen shot 2017-06-01 at 15 07 16](https://cloud.githubusercontent.com/assets/11613312/26681086/257087c2-46dc-11e7-84ea-fcf1f5c2dc69.png)

## Context
RANSAC estimation appears to be useless for many real-world datasets

**Comment by taketwo** [comment]
Just for the record, few days ago I fitted planes with RANSAC to points clouds of 1M+ points and it worked. Of course, this does not mean that there are no bugs in the implementation.

**Comment by themightyoarfish** [comment]
Scaling the cloud with a factor of ten and then doing the fitting seems to work. From that I conclude it is a problem of the density and perhaps small distances among the points give rise to numerical instabilities.

In the forum there is [one thread](http://www.pcl-users.org/Problems-with-ransac-plane-segmentation-td4033311.html) without a real solution.

**Comment by taketwo** [comment]
Interesting observation. If there is a numerical problem, then it must be in one of [these functions](https://github.com/PointCloudLibrary/pcl/blob/master/sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp) for model fitting and distance calculation.

**Comment by themightyoarfish** [comment]
Maybe it could be this part: https://github.com/PointCloudLibrary/pcl/blob/master/sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp#L95

Subtraction is being done there which could underflow if the coordinate values are very close to equal.

**Comment by taketwo** [comment]
What about trying double-precision for `p1p0` and `p2p0`? Can you give it a try?

**Comment by t-lou** [comment]
Hi, I think I know the problem. The points are quit far away from the original point, making some components of position(X, Y or Z, should be X in this data) have too much weight. A simple solution could be to transform the point cloud so that the centroid is (0,0,0). I can share the code if you need. Of course we need to compensate the transform with simple geometry. 

Result:
Without transform: [0.0351476, 0.999373, 0.00433326, 0.0229562]
With transform: [0.999295, -0.0368609, -0.00712418, 0.682014]
Looks correct according to your figures. I need to ssh to lab computer, so just printed the parameters.

**Comment by themightyoarfish** [comment]
@taketwo I'll need to see when I get to that, but I'll try at some point.

@t-lou Interesting, but the distance from the origin can't be the problem, otherwise why would scaling and rescaling wor

**Comment by t-lou** [comment]
Whatever, it seems to work here. Tell me when you need an example of code. I once tried to make a pull request, but then I realized it is just 20 lines in real application

**Comment by stefanbuettner** [comment]
Hi there,

@themightyoarfish I played a bit with the example code you provided (:+1:) and disabling model optimization yields correct results. This behavior is also observed in [the thread you referenced](http://www.pcl-users.org/Problems-with-ransac-plane-segmentation-td4033311.html) although they don't follow this observation. Personally, I think the optimization might be the problem because the covariance matrix of the plane inliers (the whole cloud in your case) and its eigen vector of the smallest eigen value is used as the plane's normal (see [SampleConsensusModelPlane::optimizeModelCoefficients](https://github.com/PointCloudLibrary/pcl/blob/master/sample_consensus/include/pcl/sample_consensus/impl/sac_model_plane.hpp#L212)). I don't remember if this was a stable method.

**Comment by SergioRAgostinho** [comment]
Guys, have a look at the discussion in #1407. It's starting sound like it's an issue arising from the single pass method.

**Comment by Jmeyer1292** [comment]
I've had serious issues in the past (Fall of last year) with the PCL 1.7 optimization on large point clouds during some work with precision laser profilers. It's not pretty, but I'll leave this code sample here because it worked well for me: It's an SVD based least squares optimization and it's much more numerically stable.
```c++
Eigen::Vector4f fitPlaneManually(const pcl::PointCloud<pcl::PointXYZ>& cloud)
{
  Eigen::MatrixXd lhs (cloud.size(), 3);
  Eigen::VectorXd rhs (cloud.size());

  for (size_t i = 0; i < cloud.size(); ++i)
  {
    const auto& pt = cloud.points[i];
    lhs(i, 0) = pt.x;
    lhs(i, 1) = pt.y;
    lhs(i, 2) = 1.0;

    rhs(i) = -1.0 * pt.z;
  }

  Eigen::Vector3d params = lhs.jacobiSvd(Eigen::ComputeThinU | Eigen::ComputeThinV).solve(rhs);

  Eigen::Vector3d normal (params(0), params(1), 1.0);
  auto length = normal.norm();

  normal /= length;
  params(2) /= length;

  return {normal(0), normal(1), normal(2), params(2)};
}
```

**Comment by mvieth** [comment]
Fixed by pull request #4983
